src/GPUActivityAgent.cpp

```diff
@@ -0,0 +1,436 @@
+                // address the issues that come from workloads have short phases that are
+                // frequency sensitive.  If a workload has a compute activity of 0.5, and
+                // is resident on the GPU for 50% of cycles (0.5) it is treated as having
+                // a 1.0 compute activity value
+                //
+                // For inactive regions the frequency selection is simply the efficient
+                // frequency from system characterization.
+                //
+                // This approach assumes the efficient frequency is suitable as both a
+                // baseline for active regions and and inactive regions. This is generally
+                // true when the efficient frequency consumes low power at idle due to clock
+                // gating or other hardware PM techniques.
+                //
+                // If f_efficient does not meet these criteria this behavior can still be
+                // achieved through tracking the GPU Utilization signal and setting frequency
+                // to a separate idle value (f_idle) during regions where GPU Utilizaiton is
+                // zero (or below some bar).
+                if (!std::isnan(gpu_utilization) &&
+                    gpu_utilization > 0) {
+                    gpu_utilization = std::min(gpu_utilization, 1.0);
+                    f_request = m_f_efficient + m_f_range * (gpu_core_activity / gpu_utilization);
+                }
+                else {
+                    f_request = m_f_efficient + m_f_range * gpu_core_activity;
+                }
+
+                // Tracking logic.  This is not needed for any performance reason,
+                // but does provide useful metrics for tracking agent behavior
+                if (gpu_core_activity >= M_GPU_ACTIVITY_CUTOFF) {
+                    m_gpu_active_region_stop.at(domain_idx) = 0;
+                    if (m_gpu_active_region_start.at(domain_idx) == 0) {
+                        m_gpu_active_region_start.at(domain_idx) = m_time.value;
+                        m_gpu_active_energy_start.at(domain_idx) = m_gpu_energy.at(domain_idx).value;
+                    }
+                }
+                else {
+                    if (m_gpu_active_region_stop.at(domain_idx) == 0) {
+                        m_gpu_active_region_stop.at(domain_idx) = m_time.value;
+                        m_gpu_active_energy_stop.at(domain_idx) = m_gpu_energy.at(domain_idx).value;
+                    }
+                }
+            }
+
+            // Frequency clamping
+            if (f_request > m_f_max || f_request < m_f_efficient) {
+                ++m_gpu_frequency_clipped;
+            }
+            f_request = std::min(f_request, m_f_max);
+            f_request = std::max(f_request, m_f_efficient);
+
+            // Store frequency request
+            gpu_freq_request.push_back(f_request);
+        }
+
+        if (!gpu_freq_request.empty()) {
```

Comment:
Why is this check necessary? Is it possible M_NUM_GPU could be 0? Will this agent load if there are no GPUs available?

Reply:
It's no longer necessary.  In previous revisions it was possible to hit a condition where the gpu_freq_request had not been initialized.  It may now be removed.